### ISSUE TITLE
Auto-switch language for spellchecking

### ISSUE THREAD

**Issue by Glitchy-Tozier** [florisboard/florisboard#1129]
In my experience, when using two or more different languages on a keyboard, most of the time I forget to switch to the correct one.

### A fun feature could be this:
If FlorisBoard recognizes that ... let's say over 80% of the typed words are incorrect, it goes through the other languages. If with one of those languages only 20% are incorrect, FlorisBoard automatically switches to the other language.

Of course this should require some safety-mechanisms like the user having typed at least a certain amount of words (5?).

**Comment by Glitchy-Tozier** [comment]
This could be useful in general, not just regarding spell-checking.

**Comment by patrickgold** [comment]
In general I see where you are heading and the idea is good, though I see several issues:

- Often times for different languages you may have different layouts. If I auto-switch the subtype because FlorisBoard detects that the other language is written, this causes the layout to change and the user to miss-type 1-3 letters immediately after the layout switch.
- Performance wise this feature may be a temporary set-back, as while searching for matches in different languages, the background CPU usage goes up and the keyboard may feel laggy.

I am a bit unsure if I add such feature or if I will rely on multi-lingual subtype support in the future.

**Comment by Glitchy-Tozier** [comment]
> Often times for different languages you may have different layouts. If I auto-switch the subtype because FlorisBoard detects that the other language is written, this causes the layout to change and the user to miss-type 1-3 letters immediately after the layout switch.

This is true, and it's something important to consider.
However, I'd like to also mention that this could be an advantage. Often times I forget to switch from English to German. This results in me getting "7" instead of "ü", "9" instead of ö", "@" instead of "ä", and so on. This could be prevented by an automatic switch.

> Performance wise this feature may be a temporary set-back, as while searching for matches in different languages, the background CPU usage goes up and the keyboard may feel laggy.

Yup, this might be an issue.


Hey, my first proposal-maybe! (I think)   :)

**Comment by Glitchy-Tozier** [comment]
(Btw, the ü/7-frustration was what spawned this idea in the first place)

**Comment by TSnatch72** [comment]
Maybe better than this would be to allow simultaneous spellchecking and suggestions for all your languages do you can write words from different languages on the same sentence without them getting flagged as wrong

**Comment by videbar** [comment]
Maybe a way of implementing this would be to have an option (maybe something called "shared dictionary for multiple languages") that would combine the dictionaries for all languages into one and use that macro-dictionary for the spellcheck. 

For example you would have a dictionary that contained all German words and all English words. 

I guess this should have less impact on performance, but it would not automatically change the keyboard layout.

**Comment by Glitchy-Tozier** [comment]
I agree that this is an interesting idea in some cases, but it doesn't fix one of my main reasons for posting this proposal:

> Often times I forget to switch from English to German. This results in me getting "7" instead of "ü", "9" instead of ö", "@" instead of "ä", and so on. This could be prevented by an automatic switch.

Basically, with a shared library, you'd still either have to choose one set of diacritics and manually switch to the other one OR make compromises with your diacritics-placements.

**Comment by patrickgold** [comment]
@Glitchy-Tozier regarding your main reason for the proposal: couldn't you specify two subtypes, one for English and one for German for instance and set them both to the same layout and popup mapping? This way layout and popup wise nothing changes, only the spell checker would show everything red until your manual switch (or automatic once I implemented it).

**Comment by Glitchy-Tozier** [comment]
> Glitchy: …OR make compromises with your diacritics-placements.

@patrickgold This would be the compromise-option.

In my specific case it would be a good idea; however, mixing languages might result in badly placed diacritics for one of those languages.

That being said, It's not too problematic, as it might feel weird to automatically have diacritics in your popups getting reordered.

**Comment by patrickgold** [comment]
You are correct, this works perfectly fine for two Latin languages such as German and English but would horrendously fail for English/Russian for instance.
